stable_radical_optimization/stable_radical_opt.py

```diff
@@ -0,0 +1,334 @@
+import argparse
+import pathlib
+import logging
+import time
+from typing import Tuple
+
+import sqlalchemy
+from sqlalchemy import create_engine
+
+import numpy as np
+import pandas as pd
+
+import rdkit
+from rdkit import Chem
+from rdkit.Chem import Mol, MolToSmiles
+
+from rlmolecule.molecule.policy import preprocessor
+
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+#sys.path.append('/projects/rlmolecule/pstjohn/models/20201031_bde/')
+#from preprocess_inputs import preprocessor as bde_preprocessor
+#bde_preprocessor.from_json('/projects/rlmolecule/pstjohn/models/20201031_bde/preprocessor.json')
+# TODO would we need a different preprocessor.json file here
+# than we have in rlmolecule.molecule.policy.preprocessor.json?
+bde_preprocessor = preprocessor.load_preprocessor()
+
+
+def construct_problem(
+        stability_model: pathlib.Path,
+        redox_model: pathlib.Path,
+        bde_model: pathlib.Path,
+        **kwargs):
+    # We have to delay all importing of tensorflow until the child processes launch,
+    # see https://github.com/tensorflow/tensorflow/issues/8220. We should be more careful about where / when we
+    # import tensorflow, especially if there's a chance we'll use tf.serving to do the policy / reward evaluations on
+    # the workers. Might require upstream changes to nfp as well.
+    from rlmolecule.tree_search.reward import RankedRewardFactory
+    from rlmolecule.molecule.molecule_config import MoleculeConfig
+    from rlmolecule.molecule.molecule_problem import MoleculeTFAlphaZeroProblem
+    from rlmolecule.molecule.molecule_state import MoleculeState
+    import tensorflow as tf
+
+
+    class StableRadOptProblem(MoleculeTFAlphaZeroProblem):
+
+        def __init__(self,
+                     engine: 'sqlalchemy.engine.Engine',
+                     config: 'MoleculeConfig',
+                     stability_model: 'tf.keras.Model',
+                     redox_model: 'tf.keras.Model',
+                     bde_model: 'tf.keras.Model',
+                     **kwargs) -> None:
+            super(StableRadOptProblem, self).__init__(engine, config, **kwargs)
+            self.engine = engine
+            self._config = config
+            self.stability_model = stability_model
+            self.redox_model = redox_model
+            self.bde_model = bde_model
+
+        def get_initial_state(self) -> MoleculeState:
+            return MoleculeState(rdkit.Chem.MolFromSmiles('C'), self._config)
+
+        def get_reward(self, state: MoleculeState) -> Tuple[float, dict]:
+            # TODO what does this do? Is it needed?
+            ## Node is outside the domain of validity
+            #elif ((node.policy_inputs['atom'] == 1).any() | 
+            #    (node.policy_inputs['bond'] == 1).any()):
+            #    return config.min_reward
```

Comment:
The preprocessor will put a '1' in an atom or bond slot if it doesn't recognize it. (0 is the padding token, and stored atoms start with '2'). So if we see that, we really shouldn't be making a prediction, since it's likely a bogus molecular structure anyways.